source/common/ur_pool_manager.hpp

```diff
@@ -0,0 +1,179 @@
+/*
+    *
+    * Copyright (C) 2023 Intel Corporation
+    *
+    * SPDX-License-Identifier: MIT
+    *
+    */
+
+#ifndef USM_POOL_MANAGER_HPP
+#define USM_POOL_MANAGER_HPP 1
+
+#include "ur_api.h"
+#include "ur_pool_manager.hpp"
+#include "ur_util.hpp"
+
+#include <functional>
+#include <vector>
+
+namespace usm {
+
+/// @brief describes an internal USM pool instance.
+struct pool_descriptor {
+    ur_usm_pool_handle_t poolHandle;
+
+    ur_context_handle_t hContext;
+    ur_device_handle_t hDevice;
+    ur_usm_type_t type;
+    bool deviceReadOnly;
+
+    static bool equal(const pool_descriptor &lhs, const pool_descriptor &rhs);
+    static std::size_t hash(const pool_descriptor &desc);
+};
+
+/// @brief creates pool descriptors for a given context, devices and pool flags.
+std::vector<pool_descriptor>
+createPoolDescriptors(ur_usm_pool_handle_t poolHandle,
+                      ur_context_handle_t hContext, uint32_t deviceCount,
+                      const ur_device_handle_t *phDevices);
+
+static inline std::pair<ur_result_t, std::vector<ur_device_handle_t>>
+urGetSubDevices(ur_device_handle_t hDevice) {
+    size_t n_compute_units;
+    auto ret =
+        urDeviceGetInfo(hDevice, UR_DEVICE_INFO_MAX_COMPUTE_UNITS,
+                        sizeof(n_compute_units), &n_compute_units, nullptr);
+    if (ret != UR_RESULT_SUCCESS) {
+        return {ret, {}};
+    }
+
+    ur_device_partition_property_t properties[] = {
+        UR_DEVICE_PARTITION_EQUALLY, static_cast<intptr_t>(n_compute_units), 0};
+
+    // Get the number of devices that will be created
+    uint32_t n_devices;
+    ret = urDevicePartition(hDevice, properties, 0, nullptr, &n_devices);
+    if (ret != UR_RESULT_SUCCESS) {
+        return {ret, {}};
+    }
+
+    std::vector<ur_device_handle_t> sub_devices(n_devices);
+    ret = urDevicePartition(hDevice, properties,
+                            static_cast<uint32_t>(sub_devices.size()),
+                            sub_devices.data(), nullptr);
+    if (ret != UR_RESULT_SUCCESS) {
+        return {ret, {}};
+    }
+
+    return {UR_RESULT_SUCCESS, sub_devices};
+}
+
+static inline bool
+isSharedAllocationReadOnlyOnDevice(const pool_descriptor &desc) {
+    return desc.type == UR_USM_TYPE_SHARED && desc.deviceReadOnly;
+}
+
+inline bool pool_descriptor::equal(const pool_descriptor &lhs,
```

Comment:
It might be useful to have some differentiation (e.g., to prefer local memory or something). And I'm not sure about implementing equals for this - I think it should be a higher-levels decision if multiple ur devices use the same pool (or use each-other separate pools).